ipld/src/selector/walk.rs

```diff
@@ -0,0 +1,208 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use super::super::{Error, Ipld, Path, PathSegment};
+use super::Selector;
+use async_recursion::async_recursion;
+use async_trait::async_trait;
+use cid::Cid;
+
+impl Selector {
+    /// Walks all nodes visited (not just matched nodes) and executes callback with progress and
+    /// Ipld node. An optional link loader/ resolver is passed in to be able to traverse links.
+    pub async fn walk_all<L, F>(
+        self,
+        ipld: &Ipld,
+        resolver: Option<L>,
```

Comment:
for what benefit? That just seems less readable

Reply:
Like, what I mean by this is when you don't specify a link resolver the type inference is either more verbose or you pass in something other than `None` into the function and that be less readable. I think usability is more important than possibly some checks removed in the code (and less prone to unexpected behaviour if expanded on)

Reply:
I guess what I'm saying is that it's double that the link resolver is optional, and the ipld it returns is also optional. if you require an `L` and use `()` or some other type instead of `None` then the code becomes simpler, I've tried it — I'm also happy to make this change later if my description isn't really clear

Reply:
Yeah but then it's not obvious that the resolver is optional and a little janky to read when you are passing () or otherwise in as a parameter. That is specifically what `Option` is for, plus also you are auto deriving the trait impl for every linkresolver impl.

Maybe I'm misunderstanding a detail of what you are saying. I am open to considering, it just seems like an opinionated and less idiomatic API

Reply:
>That is specifically what `Option` is for

I would definitely agree in the case of a concrete type; the annoying thing about `Option` in combination with generics though is that the `None` is still generic over some `T` that you don't really care about when you don't want to pass a link resolver at all.

Anyway, this whole API is inherently not very idiomatic Rust because of how directly it was ported from Go, so it's not a big deal whether this part is idiomatic or not. This is definitely not a blocker, but rather something to think about when we decide to refactor it.